pkg/backup/backupdest/backup_index.go

```diff
@@ -123,6 +125,145 @@ func IndexExists(ctx context.Context, store cloud.ExternalStorage, subdir string
 	return indexExists, nil
 }
 
+// ListIndexes lists all the index files for a backup chain rooted by the full
+// backup indicated by the subdir. The store should be rooted at the default
+// collection URI (the one that contains the `index/` directory). It returns
+// the basenames of the listed index files. It assumes that the subdir is
+// resolved and not `LATEST`.
+//
+// The indexes are returned in reverse chronological order, i.e. the most recent
+// index is first in the list.
+func ListIndexes(
+	ctx context.Context, store cloud.ExternalStorage, subdir string,
+) ([]string, error) {
+	var indexBasenames []string
+	if err := store.List(
+		ctx,
+		indexSubdir(subdir)+"/",
+		"",
+		func(file string) error {
+			indexBasenames = append(indexBasenames, path.Base(file))
+			return nil
+		},
+	); err != nil {
+		return nil, errors.Wrapf(err, "listing indexes in %s", subdir)
+	}
+	return indexBasenames, nil
+}
+
+// GetBackupTreeIndexMetadata concurrently retrieves the index metadata for all
+// backups within the specified subdir, up to the specified end time, inclusive.
+// It also returns the total size of memory allocated to store the index files.
+// The returned indexes are sorted in reverse chronological order, i.e. the
+// latest index is first in the list. The store should be rooted at the
+// collection URI that contains the `index/` directory.
+func GetBackupTreeIndexMetadata(
+	ctx context.Context, store cloud.ExternalStorage, subdir string, endTime hlc.Timestamp,
+) ([]backuppb.BackupIndexMetadata, error) {
+	indexBasenames, err := ListIndexes(ctx, store, subdir)
+	if err != nil {
+		return nil, err
+	}
+
+	// Note that the following logic depends on the fact that ListIndexes
+	// returns the index files in reverse chronological order.
```

Comment:
new comment: could you explain the high level approach here? I think we are:
- cropping the list on startIdx (i.e. removing inc backups that are newer than endtime), by _reverse_ iterating through the index names
- then, cropping it again on startIdx, by _forward_ iterating through the metadata

this feels quite complex. could we simply do one forward iteration on the index file names to get the same results? Happy to hop on a chat if easier.


old comment: nit: i would state plainly the logic, perhaps with an example. E.g. "first, reverse iterate through the chain to find the first backup with an endtime _greater_ than the specified end time
```
// endtime: 6
// (8,10), (6,8),(4,6) (2,4)
//  gets idx for (8,6)
```
"

 it took me while to wrap my head around this.